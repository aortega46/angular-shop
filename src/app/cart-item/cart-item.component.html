<li class="border-b-slate-500 pb-16">
  <img
    class="aspect-video w-full"
    src="{{ product.thumbnail }}"
    alt="{{ product.title }}" />
  <div>
    <strong>{{ product.title }}</strong> - {{ product.price }}
  </div>
  <footer class="flex items-center justify-center gap-2">
    <small>Qty: {{ product.quantity }}</small>
    <button class="p-2 hover:scale-110">+</button>
  </footer>
</li>
