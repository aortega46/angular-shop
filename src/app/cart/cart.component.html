<button
  class="absolute right-2 top-2 z-50 flex h-12 w-12 cursor-pointer items-center justify-center rounded-full bg-blue-400 p-1"
  (click)="toggleCart()">
  <mat-icon class="material-icons-outlined" fontIcon="shopping_cart"></mat-icon>
</button>

<aside
  class=" bg-black p-8 top-0 right-0 transition {{
    isCartOpened
      ? 'translate-x-0 fixed h-full w-52'
      : 'translate-x-full fixed h-full'
  }}">
  <ul>
    @for (product of []; track $index) {
      <cart-item [product]="product" />
    }
  </ul>

  <button>
    <mat-icon
      class="material-icons-outlined"
      fontIcon="remove_shopping_cart"></mat-icon>
  </button>
</aside>
