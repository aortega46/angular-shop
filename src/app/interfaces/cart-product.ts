export interface CartProduct {
  id: number
  title: string
  price: number
  thumbnail: string
  quantity: number
}
