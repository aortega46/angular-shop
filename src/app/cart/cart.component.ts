import { Component } from '@angular/core'
import { MatIconModule } from '@angular/material/icon'
import { CartItemComponent } from '../cart-item/cart-item.component'

@Component({
  selector: 'app-cart',
  standalone: true,
  imports: [MatIconModule, CartItemComponent],
  templateUrl: './cart.component.html',
})
export class CartComponent {
  isCartOpened: boolean = false

  toggleCart() {
    this.isCartOpened = !this.isCartOpened
  }
}
