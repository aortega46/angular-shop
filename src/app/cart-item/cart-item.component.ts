import { Component, EventEmitter, Input, Output } from '@angular/core'
import { MatIconModule } from '@angular/material/icon'
import { CartProduct } from '../interfaces/cart-product'

@Component({
  selector: 'cart-item',
  standalone: true,
  imports: [MatIconModule],
  templateUrl: './cart-item.component.html',
})
export class CartItemComponent {
  @Input() product!: CartProduct

  @Output() onAddToCart = new EventEmitter<CartProduct>()

  addToCart() {
    this.onAddToCart.emit(this.product)
  }
}
