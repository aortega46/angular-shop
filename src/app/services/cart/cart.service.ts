import { Injectable } from '@angular/core'
import { Product } from '../../interfaces/product'
import { BehaviorSubject, Observable, of, switchMap } from 'rxjs'
import { CartProduct } from '../../interfaces/cart-product'

@Injectable({
  providedIn: 'root',
})
export class CartService {
  cart = new BehaviorSubject<CartProduct[]>([])

  addToCart(product: Product) {
    const currentCart = this.cart.getValue()
    const productInCartIndex = currentCart.findIndex(
      item => item.id === product.id
    )

    if (productInCartIndex >= 0) {
      const newCart = [
        ...currentCart.slice(0, productInCartIndex),
        {
          ...currentCart[productInCartIndex],
          quantity: currentCart[productInCartIndex].quantity + 1,
        },
        ...currentCart.slice(productInCartIndex + 1),
      ]

      return this.cart.next(newCart)
    }

    const newCart = [...currentCart, { ...product, quantity: 1 }]
    return this.cart.next(newCart)
  }

  checkProductInCart(id: number): Observable<boolean> {
    return this.cart.pipe(
      switchMap(cart => of(cart.some(item => item.id === id)))
    )
  }

  removeFromCart({ id }: { id: number }) {
    const currentCart = this.cart.getValue()
    const newCart = currentCart.filter(item => item.id !== id)
    this.cart.next(newCart)
  }

  clearCart() {
    this.cart.next([])
  }
}
