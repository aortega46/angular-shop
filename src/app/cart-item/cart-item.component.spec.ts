import { ComponentFixture, TestBed } from '@angular/core/testing'

import { CartItemComponent } from './cart-item.component'
import { CartProduct } from '../interfaces/cart-product'
import { CartComponent } from '../cart/cart.component'

const productMock: CartProduct = {
  id: 1,
  title: 'Test',
  price: 100,
  thumbnail: 'test.png',
  quantity: 1,
}

describe('CartItemComponent', () => {
  let component: CartItemComponent
  let fixture: ComponentFixture<CartItemComponent>
  let compiled: HTMLElement

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [CartItemComponent],
      providers: [
        {
          provide: CartComponent,
          useValue: { addToCart: (product: CartProduct) => product.quantity++ },
        },
      ],
    }).compileComponents()

    fixture = TestBed.createComponent(CartItemComponent)
    component = fixture.componentInstance

    component.product = productMock

    fixture.detectChanges()

    compiled = fixture.nativeElement
  })

  it('should create', () => {
    expect(component).toBeTruthy()
  })

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot()
  })

  it('should emit when #addToCart', () => {
    const spyOnAddToCart = jest.spyOn(component.onAddToCart, 'emit')

    component.addToCart()

    expect(spyOnAddToCart).toHaveBeenCalledWith(component.product)
  })

  it('should update cart item when #addToCart', () => {
    const spyOnAddToCart = jest.spyOn(component.onAddToCart, 'emit')
    const cart = TestBed.inject(CartComponent)

    spyOnAddToCart.mockImplementation(prod => cart.addToCart(prod!))

    component.addToCart()

    expect(spyOnAddToCart).toHaveBeenCalledWith(component.product)
    expect(component.product.quantity).toBe(2)
  })
})
