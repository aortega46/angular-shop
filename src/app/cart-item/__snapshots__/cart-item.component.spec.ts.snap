// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItemComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <li
    class="border-b-slate-500 pb-4"
  >
    <img
      alt="Test"
      class="aspect-video w-full rounded"
      src="test.png"
    />
    <div
      class="m-2"
    >
      <strong>
        Test
      </strong>
       - $100 
    </div>
    <footer
      class="flex items-center justify-center gap-2"
    >
      <small>
        Qty: 1
      </small>
      <button
        class="rounded bg-slate-800 px-3 hover:outline hover:outline-1 hover:outline-blue-400"
      >
         + 
      </button>
    </footer>
  </li>
</div>
`;
